doc/adr/0002-global-and-state-channel-snapshots.md

```diff
@@ -0,0 +1,37 @@
+# 2. Global and StateChannel snapshots
+
+Date: 2022-01-05
+
+## Status
+
+Proposed
+
+## Context
+
+Global snapshots and StateChannel snapshots should be defined with their
+responsibilities.
+
+## Decision
+
+### StateChannel snapshots
+
+L0 Cell can be identified as a proxy point between Layer 1 and Layer 0. Each
+state channel will have its own L0 Cell registered in the Layer 0. One of the
+responsibilities of the L0 Cell is to create a StateChannel snapshot. It can be
+performed either by proxying the snapshot received from Layer 1 or by
+periodically creating StateChannel snapshots inside the L0 Cell. Overall, the
+StateChannel snapshot needs to be returned as an output of an L0 Cell so then it
+will be aggregated by Layer 0 and proposed to the Global snapshot.
+
+### Global snapshots
+
+Global snapshots are being created on Layer 0. They contains both DAG blocks and
+StateChannel snapshots. Trigger for the global snapshot will be described by its
+own ADR. Global snapshots that passed majority algorithm can be considered final
+and immutable aka. source of truth.
```

Comment:
I think there's potentially some ambiguity in this statement. Snapshots are the best way to organize large volumes of data for the long-term, but are not ideal as the only 'source of truth' as it introduces too many potential sources of delays. Any global value supporting machines on different configurations will hit the same problems as many other single-hash bound chains (with large networks and uncertain computing power.) The main value in a trust based partitioning structure / parallelism is attempting to offer finality in advance of a single hash merger (i.e. a Merkle root of the corresponding snapshot block.) This statement is unconstrained as well (does it require constraints on the snapshot node minimum sizes / latency?) Also -- does that mean that blocks formed in advance of the snapshot can never be considered 'finalized'? It implies that we're degenerate down to a single chain solution with the intermediary structure being essentially an ephemeral solution -- which is not ideal. 

Ideally, finality protections should be designed around preventing excessive forks in the event of legitimate confusion on the network (i.e. dropped packets, latency, other determinism issues.) The cut-off point for 'finality' should be based around nodes attempting to follow the rules as opposed to the data hierarchy. Most conflicts are immediately resolvable through some level of determinism based on trust or chain data or some trivial resolution criteria (with regards to two double spend transactions that are relatively close in time -- it's irrelevant which is actually accepted.) The main purpose of finality here is to prevent the situation of a double spend happening at distinctly different points in time (one much later than another.) Relying purely on the global snapshot phenomena to stop the second case imposes arbitrary constraints on the overall network architecture -- and so finality should be ensured well in advance of that.

Ideally, a snapshot is really only calculating a global view onto data that already exists across the network -- well after consensus has been achieved. The main benefits of it are for calculating diffs more efficiently to provide a foundational data structure.

Reply:
Ideally each node should have the same contents in its 'view' or snapshot proposal, but that's really more of an optimization. We still need to be able to handle forks in a global hierarchy to perform the most important task of the protocol which is preventing double spends. At the end of the day, we want the network to fork and potentially stall even, if an invalid/double spend block is committed to a local chain; network partitions and ddos should be optimized using reputation and health checks to prevent massive O(n) redownloads.

To your point, the goal of snapshots are to form a linear data structure (out pf path connected blocks) concurrently, while L1 consensus commits data in parallel, forming the dag. Note also that redownloads only send a diff of missing data, which in general if properly optimized should be empty; so L0 is reduces to a concurrent subprocess.

The main optimizations that achieve the optimizations about finality are:

- What we were talking about yesterday using the rep model to identify conflicts/tips connected to double spend are the optimizations which should achieve finality locally without any holdups at the L0.
- A deterministic partitioning structure for L1 facilitator selection optimizes against diffs in the address/balance ledger and blocks/tips/double spends by federating addresses to facilitators.
- The linear address structure prevents against double spends at larger time gaps
- Reducing the dag dimensions (depth, width height) and height diff for snapshot creation reduces finallity overall